### ISSUE TITLE
Generate error for invalid curly quantifier in regex

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#12539]
Migrated from [rt.perl.org#115652](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=115652) (status was 'open')

Searchable as RT115652$

**Comment by p5pRT** [comment]
### From jrw32982@yahoo.com

#### Created by jrw32982@yahoo.com



Enhancement request\.

Please tighten up the syntax for regexes\, perhaps by adding 
&nbsp;   use feature 'strict\_quantifiers'
&nbsp;    
The effect would be that the presence of left curly in a regex would be 
valid only if it was used in a syntactically correct quantifier 
\(i\.e\. \{n\}\, \{n\,\}\, \{n\,m\}\)\.  Of course\, this shouldn't affect the use
of various interpolations\, such as $hash\{key\}\, into a regex\.

Currently\, if the use of a left curly doesn't match a valid quantifier
pattern\, the left curly is treated as a literal character\.  I just spent 
several minutes puzzling my way through the output of a program which had this
regex&#8203;: /\\G\(\.\{1\,$max\)/ until I finally noticed the mismatched left curly\.  When
I finally found it I was quite surprised that the unbalanced curly did not 
generate an error\.  After the fact\, when I checked the perlre doc\, I see that 
this behavior is documented\, but I can't imagine why it would be a good idea 
to make the syntax and semantics of curlies so loose\.

Of course\, any such change would not be backwards compatible\, so maybe a
feature pragma would be the best way to add it into new code\.  I wonder how
many other similar gotchas there are in the definition of the way the regex
engine works\.  Maybe other "misfeatures" could/should be tightened down at
the same time?

Thanks\!

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

This perlbug was built using Perl v5.6.1 - Wed May  2 17:14:53 DFT 2001
It is being executed now by  Perl v5.8.8 - Tue Jun  2 16:08:32 PAKDT 2009.

Site configuration information for perl v5.8.8:

Configured by root at Tue Jun  2 16:08:32 PAKDT 2009.

Summary of my perl5 (revision 5 version 8 subversion 8) configuration:
  Platform:
    osname=aix, osvers=5.3.0.0, archname=aix-thread-multi
    uname='aix akash79 3 5 00011a85d600 '
    config_args='-desr -Dinstallprefix=/usr/opt/perl5 -Dprefix=/usr/opt/perl5 -Dcc=xlc_r -Duseshrplib -Dusethreads'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc_r', ccflags ='-D_ALL_SOURCE -D_ANSI_C_SOURCE -D_POSIX_SOURCE -qmaxmem=-1 -qnoansialias -DUSE_NATIVE_DLOPEN -DNEED_PTHREAD_INIT -q32 -D_LARGE_FILES -qlonglong',
    optimize='-O',
    cppflags='-D_ALL_SOURCE -D_ANSI_C_SOURCE -D_POSIX_SOURCE -qmaxmem=-1 -qnoansialias -DUSE_NATIVE_DLOPEN -DNEED_PTHREAD_INIT'
    ccversion='9.0.0.2', gccversion='', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=4321
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=8
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='ld', ldflags =' -brtl -bdynamic -b32'
    libpth=/lib /usr/lib /usr/ccs/lib
    libs=-lbind -lnsl -lgdbm -ldbm -ldb -ldl -lld -lm -lcrypt -lpthreads -lc -lbsd
    perllibs=-lbind -lnsl -ldl -lld -lm -lcrypt -lpthreads -lc -lbsd
    libc=, so=a, useshrplib=true, libperl=libperl.a
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_aix.xs, dlext=so, d_dlsymun=undef, ccdlflags='  -bE:/usr/opt/perl5/lib/5.8.8/aix-thread-multi/CORE/perl.exp'
    cccdlflags=' ', lddlflags='-bhalt:4 -bexpall -G -bnoentry -lpthreads -lc'

Locally applied patches:
    


@INC for perl v5.8.8:
    /usr/opt/perl5/lib/5.8.8/aix-thread-multi
    /usr/opt/perl5/lib/5.8.8
    /usr/opt/perl5/lib/site_perl/5.8.8/aix-thread-multi
    /usr/opt/perl5/lib/site_perl/5.8.8
    /usr/opt/perl5/lib/site_perl
    .


Environment for perl v5.8.8:
    HOME=/home/jrw32982
    LANG=C
    LANGUAGE (unset)
    LC_ALL=C
    LC__FASTMSG=true
    LD_LIBRARY_PATH (unset)
    LIBPATH (unset)
    LOGDIR (unset)
    PATH=/home/jrw32982/std/bin:/home/jrw32982/sh:/home/jrw32982/bin:/home/jrw32982/vim72/bin:/usr/local/bin:/bin:/usr/bin:/usr/ucb:/usr/bin/X11:/etc:/sbin:/usr/sbin:/home/jrw32982/local:.
    PERL_BADLANG (unset)
    SHELL=/usr/bin/ksh

```
</details>

**Comment by p5pRT** [comment]
### From @jkeenan

On Fri Nov 09 09&#8203;:01&#8203;:52 2012\, jrw32982@&#8203;yahoo\.com wrote&#8203;:
> 
> This is a bug report for perl from jrw32982@&#8203;yahoo\.com\,
> generated with the help of perlbug 1\.33 running under perl v5\.8\.8\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please enter your report here\]
> 
> Enhancement request\.
> 
> Please tighten up the syntax for regexes\, perhaps by adding
>    use feature 'strict\_quantifiers'
> 
> The effect would be that the presence of left curly in a regex would
> be
> valid only if it was used in a syntactically correct quantifier
> \(i\.e\. \{n\}\, \{n\,\}\, \{n\,m\}\)\.  Of course\, this shouldn't affect the use
> of various interpolations\, such as $hash\{key\}\, into a regex\.
> 
> Currently\, if the use of a left curly doesn't match a valid quantifier
> pattern\, the left curly is treated as a literal character\.  I just
> spent
> several minutes puzzling my way through the output of a program which
> had this
> regex&#8203;: /\\G\(\.\{1\,$max\)/ until I finally noticed the mismatched left
> curly\.  When
> I finally found it I was quite surprised that the unbalanced curly did
> not
> generate an error\.  After the fact\, when I checked the perlre doc\, I
> see that
> this behavior is documented\, but I can't imagine why it would be a
> good idea
> to make the syntax and semantics of curlies so loose\.
> 

Disclaimer&#8203;:  I am not a regex\-guts expert; just another regex user\.

That being said\, I doubt that the benefit which Perl users might gain
from this requested feature would be worth the effort needed to
implement and \-\- more importantly \-\- maintain the functionality\.  As I
noted in the original RT \(see Links\)\, there's no error to report here\. 
 
> Of course\, any such change would not be backwards compatible\, so maybe
> a
> feature pragma would be the best way to add it into new code\.  I
> wonder how
> many other similar gotchas there are in the definition of the way the
> regex
> engine works\.  Maybe other "misfeatures" could/should be tightened
> down at
> the same time?
> 

I'll leave it to others to determine whether we would need to create a
new feature pragma for this\.  But the fact that we even have to consider
this suggests to me that the implementation/maintenance burden will
indeed be high\.  What you are suggesting seems to be&#8203;: "Enforce a
compile\-time check solely to permit me to receive warnings about
something the docs already warned me about\."

Regexes are not easy\.  They do require study and practice\.  And regex
users benefit strongly from discussion on mailing lists such as
perl\-help or on sites like Perlmonks\.  But I don't think they require
additional warnings or pragmata\.

So \-1 from me \.\.\. but this will inevitably a community discussion\.

Thank you for posting the RT\.

Jim Keenan

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @ap

Please refer to
http&#8203;://www\.nntp\.perl\.org/group/perl\.perl5\.porters/;msgid=50A5AF43\.4040109@&#8203;khwilliamson\.com

**Comment by p5pRT** [comment]
### From @jkeenan

On Wed Nov 21 02&#8203;:00&#8203;:20 2012\, aristotle wrote&#8203;:
> Please refer to
> http&#8203;://www\.nntp\.perl\.org/group/perl\.perl5\.porters/;msgid=50A5AF43\.4040109@&#8203;khwilliamson\.com


Aristotle\, Karl&#8203;:

Can we get an update on the status of this ticket?

Thank you very much\.
Jim Keenan

**Comment by p5pRT** [comment]
### From @ap

\* James E Keenan via RT \<perlbug\-followup@&#8203;perl\.org> \[2013\-12\-14 01&#8203;:25\]&#8203;:
> On Wed Nov 21 02&#8203;:00&#8203;:20 2012\, aristotle wrote&#8203;:
> > Please refer to
> > http&#8203;://www\.nntp\.perl\.org/group/perl\.perl5\.porters/;msgid=50A5AF43\.4040109@&#8203;khwilliamson\.com
>
> Aristotle\, Karl&#8203;:
>
> Can we get an update on the status of this ticket?

Purely in terms of bug tracker concerns I \*believe\* that this ticket is
redundant\, i\.e\. the same issue has to have been filed elsewhere as well\.
But I’m not sure\. It definitely has already been discussed on p5p and
has already been attempted – per the link I give\, at minimum\. As far as
I can tell about the effort as such\, it is currently stalled\, pending an
idea for how to cut the knot\, but the desire to make it happen remains\.

**Comment by p5pRT** [comment]
### From @khwilliamson

On 12/13/2013 09&#8203;:37 PM\, Aristotle Pagaltzis wrote&#8203;:
> \* James E Keenan via RT \<perlbug\-followup@&#8203;perl\.org> \[2013\-12\-14 01&#8203;:25\]&#8203;:
>> On Wed Nov 21 02&#8203;:00&#8203;:20 2012\, aristotle wrote&#8203;:
>>> Please refer to
>>> http&#8203;://www\.nntp\.perl\.org/group/perl\.perl5\.porters/;msgid=50A5AF43\.4040109@&#8203;khwilliamson\.com
>>
>> Aristotle\, Karl&#8203;:
>>
>> Can we get an update on the status of this ticket?
>
> Purely in terms of bug tracker concerns I \*believe\* that this ticket is
> redundant\, i\.e\. the same issue has to have been filed elsewhere as well\.
> But I’m not sure\. It definitely has already been discussed on p5p and
> has already been attempted – per the link I give\, at minimum\. As far as
> I can tell about the effort as such\, it is currently stalled\, pending an
> idea for how to cut the knot\, but the desire to make it happen remains\.
>

It is not stalled\.  What happened in 5\.18 is that we now have


---
=item Useless use of '\\'; doesn't escape metacharacter '%c'

\(D deprecated\) You wrote a regular expression pattern something like
one of these&#8203;:

&nbsp;  m\{ \\x\\\{FF\\\} \}x
&nbsp;  m\{foo\\\{1\,3\\\}\}
&nbsp;  qr\(foo\\\(bar\\\)\)
&nbsp;  s\[foo\\\[a\-z\\\]bar\]\[baz\]

The interior braces\, square brackets\, and parentheses are treated as
metacharacters even though they are backslashed; instead write&#8203;:

&nbsp;  m\{ \\x\{FF\} \}x


---
This message will be in 5\.18 and 5\.20 so that everyone relying on the 
old behavior will be warned to convert\.  \(This message in 5\.18 has found 
bugs in existing code\.\)

Basically\, the backslashes don't have any effect currently\.  This 
deprecation will allow us in 5\.22 to change so that the backslashes do 
have an effect\, and that will allow us to have m\{\.\.\.\\\{\.\.\.\} which has a 
literal backslash\.  That will in turn allow us to deprecate unescaped 
braces as literals\, which will allow us in 5\.26 to have all unescaped 
braces be metacharacters\, which will then allow us to fix this ticket 
and allow spaces and other extensions in quantifiers\.  It's a slow 
process\, but it had to start somewhere; and it has\.

**Comment by p5pRT** [comment]
### From @khwilliamson

This is actually already fixed according to the OPs wishes\.  But I'm not going to resolve it until the final step is made\, some releases in the future\.  The use of such a left brace\, as in the example in the ticket\, now raises a deprecation warning\.  I'll wait to resolve this until that warning becomes fatal\, rather than write a new ticket for that\.
\-\- 
Karl Williamson

**Comment by khwilliamson** [comment]
I believe that a warning is now generated for all occurrences where a left brace is taken as a literal when it might not have been meant to be.  That means that if it immediately follows a parenthesis, no warning is generated, as it can only ever be a literal there.  There is nothing to quantify.  Raising the warning for these contexts broke too much code, so we had to back off.
